package idgs.store.pb;

import "idgs/pb/expr.proto";

option java_outer_classname = "PbMetadata";

enum FieldLabel {
  LABEL_OPTIONAL = 1;
  LABEL_REQUIRED = 2;
  LABEL_REPEATED = 3;
}

message Field {
  required FieldLabel label 	 = 1;
  required uint32 number         = 2;
  required string name           = 3;
  required idgs.pb.DataType type = 4;
  optional string type_name      = 5;
  optional string default_value  = 6;
}

message EnumOption {
  required string enum_desc   = 1;
  required uint32 enum_number = 2;
}

message Enum {
  required string enum_name  = 1;
  repeated EnumOption option = 2;
}

message Metadata {
  required string type_name = 1;
  repeated Field field      = 2;
}

message MetadataPair {
  required Metadata key   = 1;
  required Metadata value = 2;
}

message MetadataFile {
  required string package_name   = 1;
  required string file_name      = 2;
  repeated Metadata message_type = 3;
  repeated Enum enum_type        = 4;
}